udfs/cloudbuild.yaml

```diff
@@ -14,19 +14,143 @@
+  # Step 1: Build all js libraries for BigQuery UDFs
+  ###########################################################
+- name: 'node'
+  id: 'step_1_build_bq_js_libs'
+  entrypoint: 'npm'
+  args:
+    - run-script
+    - build-all-libs
+  dir: js_libs
+  waitFor:
+    - 'step_0_install_npm_packages'
+  ###########################################################
+  # Step 2: Copy libs to GCS bucket
+  ###########################################################
 - name: gcr.io/google.com/cloudsdktool/cloud-sdk
-  entrypoint: 'bash'
-  args: ['release/build.sh', '$BRANCH_NAME', '$_PR_NUMBER']
+  id: 'step_2_copy_js_to_gcs'
+  entrypoint: 'gsutil'
+  args:
+    - cp
+    - builds/*
+    - ${_JS_BUCKET}
+  dir: js_libs
+  waitFor:
+    - 'step_1_build_bq_js_libs'
+  ###########################################################
+  # Step 3: Create python virtual environment
+  ###########################################################
+- name: python:3.8.0-slim
+  id: 'step_3_create_python_venv'
+  entrypoint: 'python3'
+  args:
+    - -m
+    - venv
+    - /workspace/venv
+  waitFor:
+    - '-'
+  ###########################################################
+  # Step 4: Install python requirements for testing
+  ###########################################################
+- name: python:3.8.0-slim
+  id: 'step_4_pip_install'
+  entrypoint: '/workspace/venv/bin/python3'
+  args:
+    - -m
+    - pip
+    - install
+    - -r
+    - tests/requirements.txt
+  waitFor:
+    - 'step_3_create_python_venv'
+  ###########################################################
+  # Step 5: Create BigQuery datasets specifically for testing
+  ###########################################################
+- name: python:3.8.0-slim
+  id: 'step_5_create_test_datasets'
+  entrypoint: '/workspace/venv/bin/python3'
+  args:
+    - tests/udf_test_utils.py
+    - --create_test_datasets
+  env:
+    - SHORT_SHA=${SHORT_SHA}
+  waitFor:
+    - 'step_4_pip_install'
+  ###########################################################
+  # Step 6: Create UDF signatures with empty bodies to prevent
+  # UDF creation errors in next step when one UDF depends
+  # on another UDF.
+  ###########################################################
+- name: python:3.8.0-slim
+  id: 'step_6_create_udf_signatures'
+  entrypoint: '/workspace/venv/bin/python3'
+  args:
+    - -m
+    - pytest
+    - --workers
+    - '100'
+    - tests/create_udf_signatures.py
+  env:
+    - SHORT_SHA=${SHORT_SHA}
+  waitFor:
+    - 'step_2_copy_js_to_gcs'
+    - 'step_5_create_test_datasets'
+  ###########################################################
+  # Step 7: Create UDFs in the test dataset from step 5
+  ###########################################################
+- name: python:3.8.0-slim
+  id: 'step_7_create_udfs'
+  entrypoint: '/workspace/venv/bin/python3'
+  args:
+    - -m
+    - pytest
+    - --workers
+    - '100'
+    - tests/test_create_udfs.py
+  env:
+    - SHORT_SHA=${SHORT_SHA}
```

Comment:
does this need `JS_BUCKET`?

Reply:
Not needed. These tests simply execute the SQL. The _JS_BUCKET variable in UDFs is replaced with the actual value by the release/build.sh script.